pwiz_tools/Shared/Common/Chemistry/Molecule.cs

```diff
@@ -17,13 +17,679 @@
  * limitations under the License.
  */
 
+using pwiz.Common.Collections;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Globalization;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Text;
+using System;
+using System.Collections;
+
 namespace pwiz.Common.Chemistry
 {
     /// <summary>
     /// Represents a parsed molecular formula.  Formulas consist of capital letters, lowercase letters
     /// and numbers.  Element names start with a capital letter.
     /// </summary>
-    public class Molecule : Formula<Molecule>
+    public class Molecule : IDictionary<string, int>
     {
+        internal int _hashCode;
+        internal int? _originalHashCode; // If this matches _hashCode, then _orderHintString should match the contents and can be used for ToString()
+        private ReadOnlyDictionary<string, int> _dict;

```

Comment:
It kind of starts with this longstanding comment:
`    public class Adduct : Immutable, IComparable, IEquatable<Adduct>, IAuditLogObject
    {
            // CONSIDER(bspratt): Nick suggests we change this ImmutableDictionary to Molecule once that is performant, and supports negative counts
                    private ImmutableDictionary<string, int> Composition { get; set; } // The chemical makeup of the adduct`

So when I started this work I looked at some profiling data. ImmutableSortedList was definitely a candidate, and indeed a little test timing program showed that lookups are literally 1000x slower that ReadOnlyDictionary (which, I believe, wasn't available back when all this was originally written).  It's no real surprise that the factory model improves on our handrolled binary search implementation.

I certainly agree with the emphasis on Sum, I'll go through and make sure I haven't missed opportunities to use that instead of Plus. So far I can see only one, in SequenceUtil.cs FormulaBuilder.

Given that Equals is abstracted, the implementation doesn't really seem like an issue in terms of robustness - either it works or it doesn't.